File: packages/image-labeling/android/src/main/kotlin/com/margelo/nitro/nitromlkit/labeling/HybridImageLabeler.kt
package com.margelo.nitro.nitromlkit.labeling

import android.graphics.Bitmap
import android.graphics.BitmapFactory
import android.net.Uri
import com.google.mlkit.vision.common.InputImage
import com.google.mlkit.vision.label.ImageLabeling
import com.google.mlkit.vision.label.defaults.ImageLabelerOptions
import com.margelo.nitro.NitroModules
import com.margelo.nitro.core.Promise
import kotlinx.coroutines.async
import kotlinx.coroutines.awaitAll
import kotlinx.coroutines.coroutineScope
import kotlinx.coroutines.tasks.await
import java.net.URL

/**
 * Native Android implementation of ImageLabeler.
 * Extends the Nitrogen-generated HybridImageLabelerSpec abstract class.
 *
 * Uses MLKit's bundled on-device image-labeling model (400+ general labels).
 * The confidenceThreshold / maxLabels options are applied in Kotlin so callers
 * can tune them per-call without rebuilding the model.
 */
class HybridImageLabeler : HybridImageLabelerSpec() {

  // A single client with the threshold floored to 0, so per-call thresholds can
  // filter down in code. Reused across calls (the model loads once).
  private val labeler by lazy {
    ImageLabeling.getClient(
      ImageLabelerOptions.Builder()
        .setConfidenceThreshold(0.0f)
        .build()
    )
  }

  // MARK: - Spec methods

  override fun label(imageUri: String, options: LabelingOptions?): Promise<Array<ImageLabel>> {
    return Promise.async {
      labelImage(
        imageUri,
        threshold = options?.confidenceThreshold ?: DEFAULT_THRESHOLD,
        maxLabels = options?.maxLabels?.toInt() ?: DEFAULT_MAX_LABELS,
      )
    }
  }

  override fun labelBatch(imageUris: Array<String>, options: BatchLabelOptions?): Promise<Array<BatchLabelResult>> {
    return Promise.async {
      val threshold = options?.confidenceThreshold ?: DEFAULT_THRESHOLD
      val maxLabels = options?.maxLabels?.toInt() ?: DEFAULT_MAX_LABELS
      val concurrency = options?.concurrency?.toInt()?.coerceAtLeast(1) ?: DEFAULT_CONCURRENCY

      val results = mutableListOf<BatchLabelResult>()
      var start = 0
      while (start < imageUris.size) {
        val end = minOf(start + concurrency, imageUris.size)
        val chunk = coroutineScope {
          (start until end).map { idx ->
            async {
              try {
                BatchLabelResult(idx.toDouble(), labelImage(imageUris[idx], threshold, maxLabels), true, null)
              } catch (e: Exception) {
                BatchLabelResult(idx.toDouble(), emptyArray(), false, e.message ?: "unknown error")
              }
            }
          }.awaitAll()
        }
        results.addAll(chunk)
        start = end
      }
      results.toTypedArray()
    }
  }

  override fun checkSafety(imageUri: String): Promise<SafetyResult> {
    return Promise.async {
      buildSafety(labelImage(imageUri, threshold = 0.0, maxLabels = Int.MAX_VALUE))
    }
  }

  override fun checkSafetyBatch(imageUris: Array<String>, options: BatchSafetyOptions?): Promise<Array<SafetyResult>> {
    return Promise.async {
      val concurrency = options?.concurrency?.toInt()?.coerceAtLeast(1) ?: DEFAULT_CONCURRENCY

      val results = mutableListOf<SafetyResult>()
      var start = 0
      while (start < imageUris.size) {
        val end = minOf(start + concurrency, imageUris.size)
        val chunk = coroutineScope {
          (start until end).map { idx ->
            async {
              try {
                buildSafety(labelImage(imageUris[idx], threshold = 0.0, maxLabels = Int.MAX_VALUE))
              } catch (e: Exception) {
                // Fail open: an unreadable image is not "unsafe".
                SafetyResult(true, emptyArray(), 1.0)
              }
            }
          }.awaitAll()
        }
        results.addAll(chunk)
        start = end
      }
      results.toTypedArray()
    }
  }

  override fun matchCategories(imageUri: String, categories: Array<String>): Promise<Array<ImageLabel>> {
    return Promise.async {
      val wanted = categories.map { it.lowercase() }.toSet()
      labelImage(imageUri, threshold = 0.0, maxLabels = Int.MAX_VALUE)
        .filter { it.text.lowercase() in wanted }
        .toTypedArray()
    }
  }

  override fun isAvailable(): Boolean = true

  // MARK: - Private helpers

  private suspend fun labelImage(uri: String, threshold: Double, maxLabels: Int): Array<ImageLabel> {
    val bitmap = loadBitmap(uri) ?: throw Error("Failed to load image: $uri")
    val image = InputImage.fromBitmap(bitmap, 0)
    val mlLabels = labeler.process(image).await()
    return mlLabels
      .filter { it.confidence >= threshold }
      .sortedByDescending { it.confidence }
      .take(maxLabels)
      .map { ImageLabel(it.text, it.confidence.toDouble(), it.index.toDouble()) }
      .toTypedArray()
  }

  private fun buildSafety(labels: Array<ImageLabel>): SafetyResult {
    // Heuristic only: MLKit's general label set is NOT a trained NSFW classifier.
    val unsafe = labels.filter { it.text.lowercase() in UNSAFE_KEYWORDS && it.confidence >= UNSAFE_MIN_CONFIDENCE }
    val maxUnsafeConf = unsafe.maxOfOrNull { it.confidence } ?: 0.0
    return SafetyResult(
      safe = unsafe.isEmpty(),
      unsafeLabels = unsafe.map { it.text }.toTypedArray(),
      safetyScore = 1.0 - maxUnsafeConf,
    )
  }

  private fun loadBitmap(uri: String): Bitmap? {
    return try {
      when {
        uri.startsWith("file://") -> BitmapFactory.decodeFile(uri.removePrefix("file://"))
        uri.startsWith("/") -> BitmapFactory.decodeFile(uri)
        uri.startsWith("content://") -> {
          val context = NitroModules.applicationContext ?: return null
          context.contentResolver.openInputStream(Uri.parse(uri))?.use { BitmapFactory.decodeStream(it) }
        }
        else -> URL(uri).openStream().use { BitmapFactory.decodeStream(it) }
      }
    } catch (e: Exception) {
      null
    }
  }

  companion object {
    private const val DEFAULT_THRESHOLD = 0.5
    private const val DEFAULT_MAX_LABELS = 10
    private const val DEFAULT_CONCURRENCY = 4
    private const val UNSAFE_MIN_CONFIDENCE = 0.5

    // Best-effort keyword heuristic for checkSafety(); see buildSafety().
    private val UNSAFE_KEYWORDS = setOf(
      "brassiere", "lingerie", "underwear", "swimwear", "bikini",
      "abdomen", "navel", "thigh", "flesh",
      "blood", "weapon", "knife", "gun", "rifle",
    )
  }
}

File: packages/image-labeling/ios/HybridImageLabeler.swift
import Foundation
import UIKit
import NitroModules
import MLKitImageLabeling
import MLKitVision

/**
 * Native iOS implementation of ImageLabeler.
 * Conforms to the Nitrogen-generated HybridImageLabelerSpec protocol.
 *
 * Uses MLKit's bundled on-device image-labeling model (400+ general labels).
 * The confidenceThreshold / maxLabels options are applied in Swift so callers
 * can tune them per-call without rebuilding the model. Mirrors the Android
 * (Kotlin) implementation.
 */
class HybridImageLabeler: HybridImageLabelerSpec {

    // MARK: - HybridObject boilerplate
    var memorySize: Int { MemoryLayout<HybridImageLabeler>.size }

    // MARK: - Defaults (match Kotlin companion object)
    private static let defaultThreshold = 0.5
    private static let defaultMaxLabels = 10
    private static let defaultConcurrency = 4
    private static let unsafeMinConfidence = 0.5

    // Best-effort keyword heuristic for checkSafety(); see buildSafety().
    private static let unsafeKeywords: Set<String> = [
        "brassiere", "lingerie", "underwear", "swimwear", "bikini",
        "abdomen", "navel", "thigh", "flesh",
        "blood", "weapon", "knife", "gun", "rifle",
    ]

    // Threshold floored to 0 so per-call thresholds can filter in code.
    // Reused across calls (the model loads once).
    private lazy var labeler: ImageLabeler = {
        let opts = ImageLabelerOptions()
        opts.confidenceThreshold = 0.0
        return ImageLabeler.imageLabeler(options: opts)
    }()

    // MARK: - Spec methods

    func label(imageUri: String, options: LabelingOptions?) throws -> Promise<[ImageLabel]> {
        return Promise.async {
            let threshold = (options?.confidenceThreshold) ?? Self.defaultThreshold
            let maxLabels = (options?.maxLabels).map { Int($0) } ?? Self.defaultMaxLabels
            return try await self.labelImage(imageUri, threshold: threshold, maxLabels: maxLabels)
        }
    }

    func labelBatch(imageUris: [String], options: BatchLabelOptions?) throws -> Promise<[BatchLabelResult]> {
        return Promise.async {
            let threshold = (options?.confidenceThreshold) ?? Self.defaultThreshold
            let maxLabels = (options?.maxLabels).map { Int($0) } ?? Self.defaultMaxLabels
            let concurrency = max(1, (options?.concurrency).map { Int($0) } ?? Self.defaultConcurrency)

            var results = [BatchLabelResult]()
            results.reserveCapacity(imageUris.count)

            for chunkStart in stride(from: 0, to: imageUris.count, by: concurrency) {
                let chunkEnd = min(chunkStart + concurrency, imageUris.count)
                let chunk = Array(imageUris[chunkStart..<chunkEnd])

                let chunkResults = await withTaskGroup(of: (Int, BatchLabelResult).self) { group in
                    for (i, uri) in chunk.enumerated() {
                        let globalIdx = chunkStart + i
                        group.addTask {
                            do {
                                let labels = try await self.labelImage(uri, threshold: threshold, maxLabels: maxLabels)
                                return (globalIdx, BatchLabelResult(index: Double(globalIdx), labels: labels, success: true, error: nil))
                            } catch {
                                return (globalIdx, BatchLabelResult(index: Double(globalIdx), labels: [], success: false, error: error.localizedDescription))
                            }
                        }
                    }
                    var out = [(Int, BatchLabelResult)]()
                    for await r in group { out.append(r) }
                    return out.sorted(by: { $0.0 < $1.0 }).map(\.1)
                }
                results.append(contentsOf: chunkResults)
            }
            return results
        }
    }

    func checkSafety(imageUri: String) throws -> Promise<SafetyResult> {
        return Promise.async {
            let labels = try await self.labelImage(imageUri, threshold: 0.0, maxLabels: Int.max)
            return self.buildSafety(labels)
        }
    }

    func checkSafetyBatch(imageUris: [String], options: BatchSafetyOptions?) throws -> Promise<[SafetyResult]> {
        return Promise.async {
            let concurrency = max(1, (options?.concurrency).map { Int($0) } ?? Self.defaultConcurrency)

            var results = [SafetyResult]()
            results.reserveCapacity(imageUris.count)

            for chunkStart in stride(from: 0, to: imageUris.count, by: concurrency) {
                let chunkEnd = min(chunkStart + concurrency, imageUris.count)
                let chunk = Array(imageUris[chunkStart..<chunkEnd])

                let chunkResults = await withTaskGroup(of: (Int, SafetyResult).self) { group in
                    for (i, uri) in chunk.enumerated() {
                        let globalIdx = chunkStart + i
                        group.addTask {
                            do {
                                let labels = try await self.labelImage(uri, threshold: 0.0, maxLabels: Int.max)
                                return (globalIdx, self.buildSafety(labels))
                            } catch {
                                // Fail open: an unreadable image is not "unsafe".
                                return (globalIdx, SafetyResult(safe: true, unsafeLabels: [], safetyScore: 1.0))
                            }
                        }
                    }
                    var out = [(Int, SafetyResult)]()
                    for await r in group { out.append(r) }
                    return out.sorted(by: { $0.0 < $1.0 }).map(\.1)
                }
                results.append(contentsOf: chunkResults)
            }
            return results
        }
    }

    func matchCategories(imageUri: String, categories: [String]) throws -> Promise<[ImageLabel]> {
        return Promise.async {
            let wanted = Set(categories.map { $0.lowercased() })
            let labels = try await self.labelImage(imageUri, threshold: 0.0, maxLabels: Int.max)
            return labels.filter { wanted.contains($0.text.lowercased()) }
        }
    }

    func isAvailable() throws -> Bool {
        return true
    }

    // MARK: - Private helpers

    private func labelImage(_ uri: String, threshold: Double, maxLabels: Int) async throws -> [ImageLabel] {
        guard let image = loadImage(from: uri) else {
            throw RuntimeError.error(withMessage: "Failed to load image: \(uri)")
        }
        let visionImage = VisionImage(image: image)
        visionImage.orientation = image.imageOrientation
        let mlLabels = try await labeler.process(visionImage)
        return mlLabels
            .filter { Double($0.confidence) >= threshold }
            .sorted { $0.confidence > $1.confidence }
            .prefix(maxLabels)
            .map { ImageLabel(text: $0.text, confidence: Double($0.confidence), index: Double($0.index)) }
    }

    private func buildSafety(_ labels: [ImageLabel]) -> SafetyResult {
        // Heuristic only: MLKit's general label set is NOT a trained NSFW classifier.
        let unsafe = labels.filter {
            Self.unsafeKeywords.contains($0.text.lowercased()) && $0.confidence >= Self.unsafeMinConfidence
        }
        let maxUnsafeConf = unsafe.map { $0.confidence }.max() ?? 0.0
        return SafetyResult(
            safe: unsafe.isEmpty,
            unsafeLabels: unsafe.map { $0.text },
            safetyScore: 1.0 - maxUnsafeConf
        )
    }

    private func loadImage(from uri: String) -> UIImage? {
        let path: String
        if uri.hasPrefix("file://") {
            path = String(uri.dropFirst(7))
        } else if uri.hasPrefix("/") {
            path = uri
        } else {
            guard let url = URL(string: uri), let data = try? Data(contentsOf: url) else { return nil }
            return UIImage(data: data)
        }
        return UIImage(contentsOfFile: path)
    }
}
